File: dotnet/src/PqcStandards/MlKem/MlKem.cs
using System.Security.Cryptography;

namespace PqcStandards.MlKem;

/// <summary>ML-KEM (FIPS 203) key encapsulation mechanism.</summary>
public static class MlKemAlgorithm
{
    /// <summary>ML-KEM.KeyGen: returns (encapsulation key ek, decapsulation key dk).</summary>
    public static (byte[] ek, byte[] dk) KeyGen(MlKemParams p)
    {
        byte[] d = RandomNumberGenerator.GetBytes(32);
        byte[] z = RandomNumberGenerator.GetBytes(32);
        return KeyGenInternal(p, d, z);
    }

    internal static (byte[] ek, byte[] dk) KeyGenInternal(MlKemParams p, byte[] d, byte[] z)
    {
        var (ekPke, dkPke) = Kpke.KeyGen(p, d);
        byte[] ekHash = HashFuncs.H(ekPke);

        // dk = (dk_pke || ek_pke || H(ek) || z)
        byte[] dk = new byte[dkPke.Length + ekPke.Length + 32 + 32];
        Buffer.BlockCopy(dkPke, 0, dk, 0, dkPke.Length);
        Buffer.BlockCopy(ekPke, 0, dk, dkPke.Length, ekPke.Length);
        Buffer.BlockCopy(ekHash, 0, dk, dkPke.Length + ekPke.Length, 32);
        Buffer.BlockCopy(z, 0, dk, dkPke.Length + ekPke.Length + 32, 32);

        return (ekPke, dk);
    }

    /// <summary>ML-KEM.Encaps: returns (shared secret K, ciphertext ct).</summary>
    public static (byte[] K, byte[] ct) Encaps(MlKemParams p, byte[] ek)
    {
        byte[] m = RandomNumberGenerator.GetBytes(32);
        return EncapsInternal(p, ek, m);
    }

    internal static (byte[] K, byte[] ct) EncapsInternal(MlKemParams p, byte[] ek, byte[] m)
    {
        byte[] ekHash = HashFuncs.H(ek);
        byte[] gInput = new byte[m.Length + ekHash.Length];
        Buffer.BlockCopy(m, 0, gInput, 0, m.Length);
        Buffer.BlockCopy(ekHash, 0, gInput, m.Length, ekHash.Length);
        byte[] g = HashFuncs.G(gInput);
        byte[] K = g[..32];
        byte[] r = g[32..];

        byte[] ct = Kpke.Encrypt(p, ek, m, r);
        return (K, ct);
    }

    /// <summary>ML-KEM.Decaps: returns shared secret K.</summary>
    public static byte[] Decaps(MlKemParams p, byte[] dk, byte[] ct)
    {
        int dkPkeLen = 384 * p.K;
        int ekLen = 384 * p.K + 32;

        byte[] dkPke = dk[..dkPkeLen];
        byte[] ekPke = dk[dkPkeLen..(dkPkeLen + ekLen)];
        byte[] h = dk[(dkPkeLen + ekLen)..(dkPkeLen + ekLen + 32)];
        byte[] z = dk[(dkPkeLen + ekLen + 32)..];

        // Decrypt to get m'
        byte[] mPrime = Kpke.Decrypt(p, dkPke, ct);

        // (K', r') = G(m' || h)
        byte[] gInput = new byte[mPrime.Length + h.Length];
        Buffer.BlockCopy(mPrime, 0, gInput, 0, mPrime.Length);
        Buffer.BlockCopy(h, 0, gInput, mPrime.Length, h.Length);
        byte[] g = HashFuncs.G(gInput);
        byte[] Kprime = g[..32];
        byte[] rPrime = g[32..];

        // Re-encrypt
        byte[] ctPrime = Kpke.Encrypt(p, ekPke, mPrime, rPrime);

        // Constant-time comparison
        bool equal = CryptographicOperations.FixedTimeEquals(ct, ctPrime);

        // Implicit rejection: J(z || ct)
        byte[] jInput = new byte[z.Length + ct.Length];
        Buffer.BlockCopy(z, 0, jInput, 0, z.Length);
        Buffer.BlockCopy(ct, 0, jInput, z.Length, ct.Length);
        byte[] Kbar = HashFuncs.J(jInput);

        // Return K' if ct matches, else Kbar (implicit rejection)
        byte[] result = new byte[32];
        for (int i = 0; i < 32; i++)
            result[i] = equal ? Kprime[i] : Kbar[i];

        return result;
    }
}

File: php/src/MlKem/MlKem.php
<?php

declare(strict_types=1);

namespace PQC\MlKem;

/**
 * ML-KEM: Module Lattice-Based Key Encapsulation Mechanism.
 * FIPS 203 Algorithms 15-17.
 */
final class MlKem
{
    /**
     * ML-KEM.KeyGen: Generate encapsulation/decapsulation key pair.
     * FIPS 203 Algorithm 15.
     *
     * @param int $level 512, 768, or 1024
     * @return array{ek: string, dk: string}
     */
    public static function keyGen(int $level): array
    {
        $d = random_bytes(32);
        $z = random_bytes(32);
        return self::keyGenInternal($d, $z, $level);
    }

    /**
     * Internal deterministic key generation (for testing).
     */
    public static function keyGenInternal(string $d, string $z, int $level): array
    {
        $kpke = Kpke::keyGen($d, $level);

        $ek = $kpke['ek'];
        $hEk = HashFuncs::H($ek);

        // dk = dk_pke || ek || H(ek) || z
        $dk = $kpke['dk'] . $ek . $hEk . $z;

        return ['ek' => $ek, 'dk' => $dk];
    }

    /**
     * ML-KEM.Encaps: Generate shared key and ciphertext.
     * FIPS 203 Algorithm 16.
     *
     * @param string $ek Encapsulation key
     * @param int $level 512, 768, or 1024
     * @return array{ct: string, ss: string} Ciphertext and 32-byte shared secret
     */
    public static function encaps(string $ek, int $level): array
    {
        $m = random_bytes(32);
        return self::encapsInternal($ek, $m, $level);
    }

    /**
     * Internal deterministic encapsulation (for testing).
     */
    public static function encapsInternal(string $ek, string $m, int $level): array
    {
        $hEk = HashFuncs::H($ek);
        $gInput = $m . $hEk;
        $gOutput = HashFuncs::G($gInput);

        $K = substr($gOutput, 0, 32);
        $r = substr($gOutput, 32, 32);

        $ct = Kpke::encrypt($ek, $m, $r, $level);

        return ['ct' => $ct, 'ss' => $K];
    }

    /**
     * ML-KEM.Decaps: Recover shared key from ciphertext.
     * FIPS 203 Algorithm 17.
     *
     * @param string $ct Ciphertext
     * @param string $dk Decapsulation key
     * @param int $level 512, 768, or 1024
     * @return string 32-byte shared secret
     */
    public static function decaps(string $ct, string $dk, int $level): string
    {
        $params = Params::get($level);
        $k = $params['k'];
        $sizes = Params::sizes($level);

        // Parse dk = dk_pke || ek || h || z
        $dkPkeLen = $sizes['dkSize'];
        $ekLen = $sizes['ekSize'];
        $dkPke = substr($dk, 0, $dkPkeLen);
        $ek = substr($dk, $dkPkeLen, $ekLen);
        $h = substr($dk, $dkPkeLen + $ekLen, 32);
        $z = substr($dk, $dkPkeLen + $ekLen + 32, 32);

        // Decrypt
        $mPrime = Kpke::decrypt($dkPke, $ct, $level);

        // Re-derive (K', r')
        $gOutput = HashFuncs::G($mPrime . $h);
        $Kprime = substr($gOutput, 0, 32);
        $rPrime = substr($gOutput, 32, 32);

        // Re-encrypt
        $ctPrime = Kpke::encrypt($ek, $mPrime, $rPrime, $level);

        // Implicit rejection: if ct != ct', return J(z || ct) instead
        $Kbar = HashFuncs::J($z . $ct);

        // Constant-time selection
        if (hash_equals($ct, $ctPrime)) {
            return $Kprime;
        }

        return $Kbar;
    }
}
